dot/core/service.go

```diff
@@ -550,3 +551,66 @@ func (s *Service) GetMetadata(bhash *common.Hash) ([]byte, error) {
 	rt.SetContextStorage(ts)
 	return rt.Metadata()
 }
+
+type (
+	// Changes represents the key-value data inside a block storage
+	Changes map[string]string
+)
+
+// QueryStorage returns the key-value data by block based on `keys` params
+// on every block starting `from` until `to` block, if `to` is not nil
+func (s *Service) QueryStorage(from *common.Hash, to *common.Hash, keys []string) (map[common.Hash]Changes, error) {
+	if from == nil {
+		return nil, errors.New("cannot query data without a starting block hash")
+	}
+
+	var err error
+	blocksToQuery := []common.Hash{*from}
+	if to != nil {
+		if blocksToQuery, err = s.blockState.SubChain(*from, *to); err != nil {
+			return nil, err
+		}
+	}
+
+	queries := make(map[common.Hash]Changes)
+	for _, hash := range blocksToQuery {
+		changes, err := s.tryQueryStorage(hash, keys)
+		if err != nil {
+			return nil, err
+		}
+
+		queries[hash] = changes
+	}
+
+	return queries, nil
+}
+
+// tryQueryStorage will try to get all the `keys` inside the block's current state
+func (s *Service) tryQueryStorage(block common.Hash, keys []string) (Changes, error) {
+	stateRootHash, err := s.storageState.GetStateRootFromBlock(&block)
+	if err != nil {
+		return nil, err
+	}
+
+	changes := make(map[string]string)
```

Comment:
Wouldn't it be better if we make this slice initially instead of a map?

Reply:
```suggestion
	changes := make(Changes)
```

Reply:
as a map, this function can return the data as `key -> value` format, I don't know if a slice will be better for represent the data inside the block